### DISCUSSION TITLE
Separate Child Gap for cousin nodes

### DISCUSSION THREAD

**Discussion by autoca** [freeplane/freeplane#672] [Help]
Hi D.

I often find myself a bit confused or distracted because nodes that are not direct siblings anymore sometimes end up being very close together visually. However, if I increase the "child gap" value (of the default format), then that helps with the not-direct-siblings but the actual siblings are then too far apart.

I am not sure what the best solution would be, though.
Why is there a "child" gab to begin with? Why is it not just a "gap" for the node itself or for all nodes in the map?

**Comment by macmarrum** [comment]
I think you just described a requirement to have all nodes use an increased gap except for leaf nodes.

You can achieve this by setting Child Gap to e.g. 20 on `Default` style, adding a new style called e.g. `Leaf children` with Child Gap 3 and adding a Map Conditional Style to set `Leaf children` when Script Filter return true.

Here's a possible Script Filter
```groovy
!node.leaf && node.children.every{ it.leaf }
// https://issues.apache.org/jira/browse/GROOVY-7207
// every() returns true for empty iterator/list
```

And here's a sample map using the proposed concept
[autocaincreased-gap-except-for-leaves.mm.zip](https://github.com/freeplane/freeplane/files/9599729/autocaincreased-gap-except-for-leaves.mm.zip)

**Comment by autoca** [comment]
Pretty neat, macmarrum, thx.
This is almost what I have in mind. Additionally to the behaviour of your conditional style, I would also like for any leaves "in between" to have a shorter distance to each other as well.

In the image that would be nodes "2.2 a" and "2.2 b". They should be closer together but NOT closer to "2.1.3" nor to "2.3.1".

(Or, as of the case in the picture, maybe they could be closer to "2.1.3" and "2.3.1" but the moment that "2.2 a" or "2.2 b" either have children of their own or become "longer" themselves, then their children or they themselves should for sure have the "bigger" distance to "2.1.3" or "2.3.1".)

![Clipboard-20220919](https://user-images.githubusercontent.com/37442404/191062733-4856bdb6-d28a-473c-9ac9-cdc4ce524ba7.png)

**Reply by macmarrum** [reply]
There is a fairly new view in Freeplane, called "Allow compact mode", which will be helpful to have `2.2 b` closed to `2.2 a`. And the action can be automated with a ChangeListener. It's an interesting challenge and I'll look into it.

**Comment by euu2021** [comment]
This is useful, indeed. I remember that I created a similar request, some years ago.

Thanks for the script, macmarrum!

Now, something that I suggest is thinking about "aparent leafs", instead of actual leafs. Because,  when the nodes are folded, they are not treated as leafs, and the result is that we don't get the same visual result as the real leafs:

![image](https://user-images.githubusercontent.com/77707706/191140517-078bd0e2-7098-4a5e-8943-3931f3cb93c3.png)

**Reply by euu2021** [reply]
I am trying to come up with the filter needed to achieve that, but my brain is not working. Basically, the final result that I want is:

- if the nodes are siblings: small gap
- if not siblings: big gap

**Reply by autoca** [reply]
As far as I see it, the gap property can only be set for all children of a node all together at once, but not for (individual) nodes themselves. Looks to me like a limitation that can't be solved right now.

**Reply by euu2021** [reply]
yes, working with the child gap will be limited.

An idea that I will test tomorrow is using a script that keeps an empty node below the siblings, in order to create a gap. So, every group of siblings would have a last sibling of the style "spacing node".

**Comment by euu2021** [comment]
Ok, I developed something that can be useful.

First, I tried adding separating nodes before and after the group of siblings. But, I noticed that adding any kind of space ends up killing the compactness, that is important to me. For example:
![image](https://user-images.githubusercontent.com/77707706/191385695-11bb77cc-dbff-41c1-90f3-92c2a1e2c2f4.png)

So, I started looking for ways to create a visual cue directly in the top and bottom siblings. I found the answer by changing the CSS:
![image](https://user-images.githubusercontent.com/77707706/191385975-9503dbc6-2697-459e-a4db-0fae759db090.png)

This creates a very subtle, non-intrusive visual cue of the limits of each group of siblings. And it's simple to use, because it's a simple conditional style set by script. For example, the script for the top sibling is this:
```
if (node == map.root)
	return false
else
	childAmount = node.parent.getChildren().size()
	if (childAmount == 1)
		return false
	else if (node.parent.getChildPosition(node) == 0)
		return true
	else	
		return false
```

There are three styles (top, bottom, and lonely child), so the visual cue can be easily changed. For example, to make it more striking (this is a crazy mix of things, just as a showcase):

![image](https://user-images.githubusercontent.com/77707706/191387072-ed5bb31f-ac1d-4489-9b91-002501bca0f2.png)

This is an example map, with the conditional styles: 
[example siblings cue.zip](https://github.com/freeplane/freeplane/files/9611989/example.siblings.cue.zip)

I just started exploring the idea. I can already think of other interesting things that can be done with it.

**Reply by autoca** [reply]
This is interesting.
One suggestion for the CSS:
Do not use `p { ... }` but instead use `body { ... }`.
When styling p in case of a multi-line node each paragraph gets that red line.

One might also try to play around with the border thickness and with the padding and margin attributes.
Nice ideas in any case!! Thank you, euu!

**Reply by autoca** [reply]
How exactly did you add those spaces in the first picture?

**Reply by euu2021** [reply]
>When styling p in case of a multi-line node each paragraph gets that red line.

I overlooked that. Thanks.

>How exactly did you add those spaces in the first picture?

I added nodes manually, just to test the visual result.

Now, there is a way to add spaces using the CSS. Put a border with the same color as the map background. It solves the problem that you mentioned above, about the middle nodes getting a big gap; and it also works when one of the children is not a leaf:

![image](https://user-images.githubusercontent.com/77707706/191628992-699c38ab-97a4-441e-8174-af0b752ad6a7.png)

On the other hand, it's not practical at all, because the bottom nodes don't work with "normal styles".
 Here is the map:
[example siblings gap.zip](https://github.com/freeplane/freeplane/files/9621105/example.siblings.gap.zip)


So, it seems that there is no good workaround for your original request. That said, I think it would be relevant to have an easy way to set different values for the gap between siblings, and the gap between not siblings. Perhaps, the solution that feeds two birds with one scone is having a cloud for siblings (currently, the cloud feature is for branches):

- it would create a margin between groups of siblings
- it would create options to better highlight groups of siblings. For example, having a color for the "ground" between each sibling, different from the background, helps in creating an intuitive sense of grouping there. Like that:

![image](https://user-images.githubusercontent.com/77707706/191630693-fe72abca-7068-43bb-941d-e18f244e2bf8.png)

**Reply by autoca** [reply]
> Now, there is a way to add spaces using the CSS. Put a border with the same color as the map background. It solves the problem that you mentioned above, about the middle nodes getting a big gap;

Thx, interesting.
But were do the black lines inside the gap come from?
![191385695-11bb77cc-dbff-41c1-90f3-92c2a1e2c2f4](https://user-images.githubusercontent.com/37442404/191740549-711af45b-4a38-4859-8483-4e3b57859686.png)

**Reply by euu2021** [reply]
They are just a node with the text `\______/`, and everything transparent, except the core text.

**Reply by aaa1386** [reply]
> Ok, I developed something that can be useful.


It is an interesting script.
The problem I have is that the red line is as long as the text I am writing. Sometimes I only write one word inside the node, so the length of the red line is small.
Is it possible to edit the script or the three existing styles so that, for example, the red line can always be 3 cm?

**Reply by euu2021** [reply]
> Is it possible to edit the script or the three existing styles so that, for example, the red line can always be 3 cm?

I don't know how to do it. It seems that the HTML element is limited to the width of the text, so the CSS styling is limited to that width.

**Reply by aaa1386** [reply]
> > Ok, I developed something that can be useful.
Whenever several paragraphs are created in a node, the styles do not work well, see the following image:
![image](https://github.com/user-attachments/assets/446a98d7-83e6-4d37-bbcc-6578155f0788)

To solve this problem, I changed the letter p in the three written styles to the word "body" or "html". Now the script and styles work well, see the following image:
![image](https://github.com/user-attachments/assets/9b3b9b51-86ee-4070-a19e-6539e484f3e3)

**Reply by aaa1386** [reply]
> > Is it possible to edit the script or the three existing styles so that, for example, the red line can always be 3 cm?
> 
> I don't know how to do it. It seems that the HTML element is limited to the width of the text, so the CSS styling is limited to that width.

Look at the image below. When I'm editing, the red line covers the entire bottom of the rectangle. Is it not possible to change the script or the three styles that are created so that the red line does not shrink after we quit editing?
![image](https://github.com/user-attachments/assets/2fc61951-0f9c-4939-b90f-3279b1cbe574)

**Reply by euu2021** [reply]
> Look at the image below. When I'm editing, the red line covers the entire bottom of the rectangle. Is it not possible to change the script or the three styles that are created so that the red line does not shrink after we quit editing?

I don't know 🤔

**Comment by awolo** [comment]
I have always been unhappy with the difficulty telling when two nodes are siblings or not, or "limits of each group of siblings". That's one of the several reasons I suggested, just a few weeks ago, "to be able to anchor the vertical position of children not to rise above parent" which has become the **Child Placement** option, right under Child Gap in the Format - Node Layout section of Styles or the side bar. Try the **Top** option. I haven't had an issue with it since.

**Reply by autoca** [reply]
nice. What is is the difference between "center" and "as parent"?

**Reply by euu2021** [reply]
I  tried it, but it doesn't help me much. It doesn't give me an intuitive idea that there are different groups of siblings there.

 Also, I prefer the classic "parent in the middle" mode, because it gets easier to jump the eyes between the siblings and the parent. When the parent gets on top, the distance to the bottom siblings quickly gets too big. I mean, if there is only one icecream stand in the beach, the most convenient placement for it is in the middle of the beach shore, so nobody needs to walk too much 😅.

>What is is the difference between "center" and "as parent"?

 "as parent" means that it will follow the same setting as the parent.

**Reply by awolo** [reply]
> only one icecream stand in the beach, 

Ah, a bittersweet example with Fall here today. Where's the beach, by the way?

For me, Outline View is helpful. Especially if you have a shortcut assigned, you can switch views back and forth quickly with the selected node staying static in the view but not necessarily the center. I have come to kind of love Outline view even though there's no mouse dragging allowed and the various gap sizes are not modifiable. Those changes would make it stellar for me.

**Reply by euu2021** [reply]
> Ah, a bittersweet example with Fall here today. Where's the beach, by the way?

Actually, the beach that I had in mind is not very exciting 🤣. I was thinking about the hypothetical beach of the Hotelling’s Game, a classic game theory problem ([explained here](https://www.youtube.com/watch?v=jILgxeNBK_8))

**Reply by awolo** [reply]
Oh, I will watch it tomorrow.
But how about Outline mode? I love switching quickly back and forth. And I'd love to see it with the enhancement ideas.

**Reply by euu2021** [reply]
Let's wait to see the improvements of the new version.

**Comment by euu2021** [comment]
Playing with the settings, I noticed something else that may help to highlight the groups of siblings. Extra thick edges:

![image](https://user-images.githubusercontent.com/77707706/191633208-11b7e05f-4708-4b78-bf50-c99af4bed681.png)

**Comment by dpolivaev** [comment]
I am impressed by the intensity of this discussion. I have implemented an extra gap for nodes with child nodes, you'll get it in the next preview.

**Reply by macmarrum** [reply]
That's great :smile: 
I'll take it for a spin once available :car:

**Reply by macmarrum** [reply]
@dpolivaev Is it in preview 1.10.5_04? 
I cannot see any extra gap for nodes with child nodes.

**Reply by euu2021** [reply]
Setting the child gap to 0, a gap can be noticed between the groups of siblings:
![image](https://user-images.githubusercontent.com/77707706/193072763-74dc4449-b543-42d5-8052-399ceb75c0fd.png)

**Reply by macmarrum** [reply]
Oh, I see.
It's a bit small to be noticeable, isn't it?

**Reply by dpolivaev** [reply]
The gap is added only if the cumulated child node height is greater than the parent node height.
It is calculated as `extraVGap = Math.max(defaultVGap, minimalDistanceBetweenChildren / 6);`where defaultVGap is a standard vertical distance between siblings and minimalDistanceBetweenChildren is a probably higher distance between siblings without children set for the current node.

Feel free to suggest how to change this rule. We can play with different variations. However, I am not going to make it configurable because the majority of users would probably never find and understand this option.

In my test map the distances look as follows:

<img width="182" alt="subtree-layout" src="https://user-images.githubusercontent.com/1558052/193130664-7277fb1d-6053-4008-85ff-f3168edbc25a.png">

**Reply by autoca** [reply]
Not sure but for my subjective taste it feels that the extra gap should be about 3 times as wide.
Still, thank you for considering it at all.

**Reply by euu2021** [reply]
>Feel free to suggest how to change this rule. We can play with different variations.

The shortcoming of your solution is the situation where there is only one child. Those kids get too close to their cousins:

![image](https://user-images.githubusercontent.com/77707706/193473253-c48077fc-89cb-438c-bec1-0d13d1bdea66.png)

I still need to understand your solution, in order to suggest another idea. The `extraVGap` is applied to what?

>However, I am not going to make it configurable because the majority of users would probably never find and understand this option.

I think it would be easy to understand if the setting is called something like "extra distance between groups of siblings". Then, use an absolute value, to be added to the value set at "child gap".

**Reply by dpolivaev** [reply]
> The shortcoming of your solution

Please be careful about using personal pronouns expressing disagreement. I wish you had written "latest solution".

The additional distance is added only if the cumulative height of the child subtrees is greater than the height of the parent node. Otherwise, the more levels you have the larger the distance and it does not feel right to me.

> I think it would be easy to understand if the setting is called something like "extra distance between groups of siblings". Then, use an absolute value, to be added to the value set at "child gap".

Freeplane currently has so many formatting options that I am hesitant of adding new ones.
Also, if added, it should be a mind map or a node property because the maps would not have similar look at different machines otherwise. I hope that the added distance is flexible enough because it automatically increases with increasing child gaps.

**Reply by euu2021** [reply]
>Otherwise, the more levels you have the larger the distance and it does not feel right to me.

I see. The problem is difficult to solve.
